### PR TITLE
[SPARK-8189] [SQL] use Long for TimestampType in SQL

### PR DESCRIPTION
**Pull request by davies**
This PR change to use Long as internal type for TimestampType for efficiency, which means it will the precision below 100ns.

### CODE REVIEW

**Comment by davies** [comment]
cc @rxin @JoshRosen

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -155,6 +156,7 @@ class CodeGenContext {
     case DoubleType => "-1.0"
     case IntegerType => "-1"
     case DateType => "-1"
+    case TimestampType => "-1L"
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: this should be 0?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala [hunk 2]
```diff
@@ -87,4 +88,33 @@ object DateUtils {
       ISO8601GMT.parse(s)
     }
   }
+
+  /**
+   * Return a java.sql.Timestamp from number of 100ns since epoch
+   */
+  def toTimestamp(ns: Long): Timestamp = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: ns sounds like nano second.

maybe do

num100ns

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala [hunk 3]
```diff
@@ -87,4 +88,33 @@ object DateUtils {
...
+    t.setNanos(nanos.toInt * 100)
+    t
+  }
+
+  /**
+   * Return the number of 100ns since epoch from java.sql.Timestamp.
+   */
+  def fromTimestamp(t: Timestamp): Long = {
+    if (t != null) {
+      t.getTime() * 10000L + (t.getNanos().toLong / 100) % 10000L
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: is this correct? I think getTime and getNanos might actually overlap in Timestamp

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: let's add unit test for this file. it is too scary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala [hunk 4]
```diff
@@ -87,4 +88,33 @@ object DateUtils {
...
+    }
+    val t = new Timestamp(seconds * 1000)
+    t.setNanos(nanos.toInt * 100)
+    t
+  }
+
+  /**
+   * Return the number of 100ns since epoch from java.sql.Timestamp.
+   */
+  def fromTimestamp(t: Timestamp): Long = {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: maybe name this fromJavaTimestamp

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala (hunk 2 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: and rename it toJavaTimestamp

File: sql/hive/src/test/resources/golden/timestamp cast #5-0-dbd7bcd167d322d6617b884c02c7f247 [hunk 5]
```diff
@@ -1 +1 @@
--0.0010000000000000009
+-0.001
```
Comment (rxin) [inline sql/hive/src/test/resources/golden/timestamp cast #5-0-dbd7bcd167d322d6617b884c02c7f247]: why did we change the output?

**Comment by rxin** [comment]
cc @liancheng to take a look at the column store stuff.

**Comment by rxin** [comment]
We also need to rewrite CastSuite to make it cover more stuff. The coverage currently is pretty low.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Use `-1` as an uninitialized value, easy for debug.

File: sql/hive/src/test/resources/golden/timestamp cast #5-0-dbd7bcd167d322d6617b884c02c7f247 (hunk 5 above)
Comment (davies) [inline sql/hive/src/test/resources/golden/timestamp cast #5-0-dbd7bcd167d322d6617b884c02c7f247]: The old result has float error.

**Comment by liancheng** [comment]
In-memory columnar part looks good. The test failure was because of timestamp precision change. We can just ignore those two Hive compatibility test cases.

**Comment by rxin** [comment]
Jenkins, retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala [hunk 6]
```diff
@@ -398,11 +396,11 @@ private[sql] object JsonRDD extends Logging {
     }
   }
 
-  private def toTimestamp(value: Any): Timestamp = {
+  private def toTimestamp(value: Any): Long = {
     value match {
-      case value: java.lang.Integer => new Timestamp(value.asInstanceOf[Int].toLong)
-      case value: java.lang.Long => new Timestamp(value)
-      case value: java.lang.String => toTimestamp(DateUtils.stringToTime(value).getTime)
+      case value: java.lang.Integer => value.asInstanceOf[Int].toLong * 10000L
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala]: why is this 10000L?

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala [hunk 7]
```diff
@@ -266,8 +267,8 @@ private[parquet] abstract class CatalystConverter extends GroupConverter {
   /**
    * Read a Timestamp value from a Parquet Int96Value
    */
-  protected[parquet] def readTimestamp(value: Binary): Timestamp = {
-    CatalystTimestampConverter.convertToTimestamp(value)
+  protected[parquet] def readTimestamp(value: Binary): Long = {
+    DateUtils.fromJavaTimestamp(CatalystTimestampConverter.convertToTimestamp(value))
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: note: we should just convert directly rather than constructing a new java timestamp

File: sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala (hunk 6 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/json/JsonRDD.scala]: milliseconds to 100ns

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala [hunk 8]
```diff
@@ -311,8 +312,9 @@ private[parquet] class RowWriteSupport extends WriteSupport[Row] with Logging {
     writer.addBinary(Binary.fromByteArray(scratchBytes, 0, numBytes))
   }
 
-  private[parquet] def writeTimestamp(ts: java.sql.Timestamp): Unit = {
-    val binaryNanoTime = CatalystTimestampConverter.convertFromTimestamp(ts)
+  private[parquet] def writeTimestamp(ts: Long): Unit = {
+    val binaryNanoTime = CatalystTimestampConverter.convertFromTimestamp(
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala]: here too: we should do the conversion directly.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala [hunk 9]
```diff
@@ -357,7 +359,7 @@ private[parquet] class MutableRowWriteSupport extends RowWriteSupport {
       case FloatType => writer.addFloat(record.getFloat(index))
       case BooleanType => writer.addBoolean(record.getBoolean(index))
       case DateType => writer.addInteger(record.getInt(index))
-      case TimestampType => writeTimestamp(record(index).asInstanceOf[java.sql.Timestamp])
+      case TimestampType => writeTimestamp(record(index).asInstanceOf[Long])
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetTableSupport.scala]: should this call getLong?

**Comment by rxin** [comment]
I'm going to merge this, but we should fix rest of the things later. 

@JoshRosen would be great if you review this again too.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala (hunk 7 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: The old code is really complicated (to be compatible with Hive), it does deserve another PR. cc @adrian-wang

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala [hunk 10]
```diff
@@ -28,6 +28,7 @@ import org.apache.spark.sql.catalyst.expressions.Cast
  */
 object DateUtils {
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: since we are supporting timestamp here, Maybe we want to rename this object to DateTimeUtils, and also update the javadoc.

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala (hunk 7 above)
Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: Yes, I'll update it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala [hunk 11]
```diff
@@ -87,4 +88,33 @@ object DateUtils {
...
+  }
+
+  /**
+   * Return the number of 100ns since epoch from java.sql.Timestamp.
+   */
+  def fromJavaTimestamp(t: Timestamp): Long = {
+    if (t != null) {
+      t.getTime() * 10000L + (t.getNanos().toLong / 100) % 10000L
+    } else {
+      0L
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateUtils.scala]: why not use -1L just like the initial value?

File: sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala (hunk 7 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: It's done here: #6759

Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/parquet/ParquetConverter.scala]: I'll help review it, Thanks!
